### ISSUE TITLE
Donate Discovery API to Knative

### ISSUE THREAD

**Issue by n3wscott** [knative/community#81]
## Objective
Scott Nichols proposes to [donate discovery](https://github.com/n3wscott/discovery) to the Knative project.

## Background

As part of trying to solve the Sources WG mission for understanding:

1. What sources are installed in my cluster?
1. What instances of those sources are running in my cluster?
   - and what is their current state?

I wrote a PoC that adds a new API Group to Kubernetes called `discovery.knative.dev` and first pass implemented a new Resource called `DuckType`. This resource in combination with another tool like [`kubectl duck`](https://github.com/n3wscott/kubectl-duck) allows me to answer the above questions.

## Current Status

The code is written to Knative standards, though needs more overall testing (more than you would do for a PoC). `discovery` was presented to the Knative Client working group mid January, 2020 and the reception was overwhelmingly positive and the client WG would like to use it as a base for understanding some of the ducktypes we have been implementing in Knative (Addressable, Subscribabe, Source, etc).

## Future Directions

This is a PoC focused around the DuckType concept, but there are other things that we would like a hook into a cluster for api-based discovery. Such as add-on locations: I know I want an addon that provides a feature, "how do I find it to install?"

Discovery might also grow to include some UI component that allows for cluster understanding, such as an evolved form of [graph](github.com/n3wscott/graph) which is heavly DuckType based.

### Initial Roadmap
* Develop out and release DuckType
* Work with Client WG on integration with DuckType
* Pitch discovery as a general solution for cluster resource understanding (bring your own ducktypes).

## Process of donation and adoption

I propose that discovery be led by the Sources WG at the start in a new repo: knative/discovery. We will focus on bringing DuckType to market and enabling complex client interactions scoped to eventing for now.

An initial OWNERS file would list the Sources WG Leads, but we expect other names will be added as contributions are accepted.

## Donation Checklist
* [x]  Original authors/owners decide a donation is appropriate
* [x]  Notification of proposal
  
  * [x]  Submit proposal to knative/community repo
  * [x]  Email notification of proposal sent to knative-users and knative-dev
* [x]  Approval of adoption by Knative committees
  
  * [ ]  Steering Committee approval on fit for the Knative project
  * [x]  Technical Oversight Committee approval to adopt the repo/code (as per its charter).
* [x]  Update code licenses
  
  * [x]  Apache License 2.0 applied to code
  * [x]  Creative Commons License 4.0 applied to documentation
* [ ]  Relocate the repository
  
  * [ ]  Create OWNERS file
  * [ ]  Create AUTHORS file
  * [ ]  Update copyright header to Copyright (C) 2019 The Knative Authors.
  * [ ]  Verify that all contributors are Google CLA signatories
  * [ ]  Adopt Google CLA bot

**Comment by matzew** [comment]
+1

**Comment by n3wscott** [comment]
Any action here?

**Comment by n3wscott** [comment]
Cc @bsnchan

**Comment by evankanderson** [comment]
What WG will own this code? Assuming that we have an ownership model in place, I'm 👍 on this proposal.

**Comment by n3wscott** [comment]
> What WG will own this code? Assuming that we have an ownership model in place, I'm 👍 on this proposal.

The sources WG would.

**Comment by evankanderson** [comment]
@lionelvillard @nachocano 

(As a heads-up on a new repo for you)

/assign @mattmoor @vaikas 

To approve for TOC as well.

I'll get someone from Steering to sign off, but as this is a component that isn't owned elsewhere and is useful to client for working with eventing, I don't think there's really a scope question.

**Comment by mattmoor** [comment]
tl;dr I am supportive of introducing a concept in knative to bring stronger typing to our duck-typed relationships, but I'm not sure this process is appropriate or the next steps.

_This came up at TOC Thursday, but the actual checklist is [here](https://github.com/knative/community/blob/master/REPOSITORY-GUIDELINES.md#requirements-for-repository-donation)._

Using the donation mechanics for this is awkward and seems like using a bazooka to kill a fly.  I like the PoC as a tool to illustrate an idea, but I'm a big believer in throwing away PoCs after using them as a tool to inform / write a design.

The nearest example to this in recent memory is what we did with Bindings and SinkBinding.   We built the [PoC](https://github.com/mattmoor/foo-binding/tree/kubecon), which informed the docs ([1](https://docs.google.com/document/d/1t5WVrj2KQZ2u5s0LvIUtfHnSonBv5Vcv8Gl2k5NXrCQ/edit?ts=5dcedafb#heading=h.lnql658xmg9p), [2](https://docs.google.com/document/d/1mBVFW5Yc_1olg-0p8gbcQDLeAETX7v_li9pZkGiGrb8/edit#heading=h.lnql658xmg9p)), and after socializing them we pulled the trigger.

My $0.02 is that the next step is a detailed design outlining what's actually proposed for what we do upstream; you might consider following the pattern of Bindings and splitting the abstract idea (Bindings) into a separate doc from the first instance (SinkBinding).  I think the docs will help paint a clearer picture of the repo requirements.

**Comment by nachocano** [comment]
The more I think about how to properly solve [this ](https://github.com/knative/eventing/issues/2484), the more I like this `discovery` API. 
Given that we are thinking of revisiting some of our [EventType registry](https://github.com/knative/eventing/issues/2750) work sometime soon, I would highly suggest to start with adding this `discovery` API to the eventing repo. Some of the work on EventType might naturally fall into this group.
We would also avoid the overhead of having to create and *maintain* a separate repo.
Once we have more clarity of the full scope, and who will be actual consumers besides eventing (e.g, CLI, projects outside Knative, etc.), then we can certainly move it out... 
I'd rather start small, iterate fast, polish things, and then upstream it if needed.

my $0.0000002

**Comment by evankanderson** [comment]
It sounds like the Eventing Sources WG (of which Scott is a lead) would like to sponsor this repo.

I'll create it later today in `knative-sandbox`; if it turns out that the Client WG would prefer to sponsor, I leave it up to them to work that out.

**Comment by mattmoor** [comment]
Likely blocked on https://github.com/knative/community/issues/152

**Comment by grantr** [comment]
Repository created in https://github.com/knative/eventing/issues/3167
